helm-v3/006-repositories.md

```diff
@@ -1,40 +1,177 @@
-and could benefit from a capability to push packages from the Helm client.
-
-To support registries there will be a `helm push` command that works in a
-similar manner to `helm fetch`. It will support different systems via the
-URI/URL scheme and provide a default implementation for HTTP(S). Other systems,
-such as S3, can implement an uploader via a plugin for schemes Helm does not
-provide an implementation for.
-
-> [name=Matt Farina] There very well may be additional work here around
-> registry search and pluggable or otherwise extendable authenication
-> mechanisms.
-> 
-> [name=Adnan Abdulhussein] What would the default HTTP(S) implementation be?
-> [name=Matt Fisher] Adnan, would you mind clarifying your comment a little?
-> 
-> [name=Adnan Abdulhussein] Sorry, the proposal mentions there will be a
-> default implemention for `helm push` to a regular HTTP(S) chart repository.
-> It's unclear to me how this would actually work.
-> 
-> [name=Adnan Abdulhussein] Sounds like this is something that will be fleshed
-> out more in Josh's proposal, so I'll wait for that :)
+- A repository v2 specification will be defined, and the index file will be
+JSON (`index.json` vs `index.yaml`)
+
+## Capabilities API
+
+The Capabilties API is designed to provide Helm with information such as:
+
+1. Does this domain support actions such as login or push?
+2. What supported method should the client use to perform a specific action?
+3. What metadata does the client need to initiate the action?
+
+This will allow Helm to support any number of things in the future related
+to repositories without making breaking changes (e.g. remote `helm search`).
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
+accepts token-based login:
+
+```
+...
+        "login": {
+            "methods": {
+                "token": {
+                    "loginUrl": "https://auth.site.com/oauth2/authorize",
+                    "tokenUrl": "https://auth.site.com/oauth2/token"
+                }
+            }
+        }
+...
+```
+
+## The `helm login` command
+
+A new `helm login` command will be added to Helm, responsible for
+managing authentication across repository providers.
+
+The basic usage will be:
+
+```
+helm login <url>
```

Comment:
what would the URL look like? Is it the root registry URL? `demo42.azurecr.io` | `docker.io` | `gcr.io/projectid`

Reply:
Will it support context path in URL?

Reply:
In the current state of this proposal, context path is not supported / ignored.

The spec on Well-Known URIs goes into detail on this:
https://tools.ietf.org/html/rfc5785

Depending on the contents of `/.well-known/helm`, it might be possible that we can enable login to one domain, which then grants access to repos hosted on other domains.

For example:
```
helm login helm.jfrog.com
helm repo add myrepo jfrog.com/helm/repos/myrepo
```

I think it depends on the contents of 401 response headers (`WWW-Authenticate`) from a protected endpoint.

Reply:
I imagine `helm login` would only work against a domain similar to `docker login`. Having to support relative URL paths as part of the login protocol would likely be fraught with peril. I like the idea of one "chartmuseum" per domain.

Reply:
the problem is different clouds implemented individual tenants differently. 
- docker.io is one domain for all users. 
- gcr.io uses one domain for all users, where the [project] is the separator (https://cloud.google.com/container-registry/docs/pushing-and-pulling) 
- aws uses inidivdual domains `docker login -u AWS -p password https://aws_account_id.dkr.ecr.us-east-1.amazonaws.com`
- quay.io - i don't fully understand their model for individual tenants (https://docs.quay.io/solution/getting-started.html)
- Azure Container Registry uses individual domains; demo42.azurecr.io, contoso.azurecr.io, ...